rpc/dialer.go

```diff
@@ -240,6 +241,152 @@ func DialDirectGRPC(ctx context.Context, address string, logger utils.ZapCompati
 }
 
+func socksProxyDialContext(ctx context.Context, network, proxyAddr, addr string) (net.Conn, error) {
+	dialer, err := proxy.SOCKS5(network, proxyAddr, nil, proxy.Direct)
+	if err != nil {
+		return nil, fmt.Errorf("error creating SOCKS proxy dialer to address %q from environment: %w",
+			proxyAddr, err)
+	}
+	conn, err := dialer.(proxy.ContextDialer).DialContext(ctx, network, addr)
+	if err != nil {
+		return nil, err
+	}
+	return conn, nil
+}
+
+// SocksProxyFallbackDialContext will return nil if SocksProxyEnvVar is not set or if trying to connect to a local address,
+// which will allow dialers to use the default DialContext.
+// If SocksProxyEnvVar is set, it will prioritize a connection made without a proxy but will fall back to a SOCKS proxy connection.
+func SocksProxyFallbackDialContext(
+	ctx context.Context,
+	network, addr string,
+	logger utils.ZapCompatibleLogger,
+) func(ctx context.Context, network, addr string) (net.Conn, error) {
+	// Use SOCKS proxy from environment as gRPC proxy dialer. Do not use SOCKS proxy if trying to connect to a local address.
+	localAddr := strings.HasPrefix(addr, "[::]") || strings.HasPrefix(addr, "localhost") || strings.HasPrefix(addr, "unix")
+	proxyAddr := os.Getenv(SocksProxyEnvVar)
+	if localAddr || proxyAddr == "" {
+		// return nil in these cases so that the default dialer gets used instead.
+		return nil
+	}
+
+	return func(ctx context.Context, network, addr string) (net.Conn, error) {
+		// if ONLY_SOCKS_PROXY specified, no need for a parallel dial - only dial through
+		// the SOCKS proxy directly.
+		if os.Getenv(OnlySocksProxyEnvVar) != "" {
+			logger.Infow("Both SOCKS_PROXY and ONLY_SOCKS_PROXY specified, only SOCKS proxy will be used for outgoing connection")
+			conn, err := socksProxyDialContext(ctx, network, proxyAddr, addr)
+			if err == nil {
+				logger.Infow("connected with SOCKS proxy")
+			}
+			return conn, err
+		}
+
+		type dialResult struct {
+			net.Conn
+			error
+			primary bool
+			done    bool
+		}
+		results := make(chan dialResult) // unbuffered
+
+		var primary, fallback dialResult
+		var wg sync.WaitGroup
+		defer wg.Wait()
+
+		// otherwise, do a parallel dial with a slight delay for the fallback option.
+		returned := make(chan struct{})
+		defer close(returned)
+
+		dialer := func(ctx context.Context, dialFunc func(context.Context) (net.Conn, error), primary bool) {
+			defer wg.Done()
+			conn, err := dialFunc(ctx)
+			select {
+			case results <- dialResult{Conn: conn, error: err, primary: primary, done: true}:
+			case <-returned:
+				if conn != nil {
+					utils.UncheckedError(conn.Close())
+				}
+			}
+		}
+
+		logger.Infow("SOCKS_PROXY specified, SOCKS proxy will be used as a fallback for outgoing connection")
+		// start the main dial attempt.
+		primaryCtx, primaryCancel := context.WithCancel(ctx)
+		defer primaryCancel()
+		wg.Add(1)
+		primaryDial := func(ctx context.Context) (net.Conn, error) {
+			var zeroDialer net.Dialer
```

Comment:
> `zeroDialer`

What's a "zero" dialer? Like the zero-valued `net.Dialer`? Just default settings?

Reply:
yep